Updates.

bequiett opened the pull request:
신규 업데이트

* 설치된 DLC 감지 추가

bequiett committed:
Updates.

New added features:
* Detect Installed DLCs
* Always on Top

Always located on top of any program including DJMAX RESPECT V.
* Can toggle this option by pressing F2
* You can only use toggle hotkey when DJMAX RESPECT V is active and focused on your screen.

bequiett committed:
Merge branch 'wowvv0w:main' into main

wowvv0w commented on DjmaxRandomSelectorV/ViewModels/MainViewModel.cs:
```diff
@@ -11,6 +11,7 @@
 using System.Windows.Interop;
 using System.Windows.Media.Effects;
 using System.IO;
+using DjmaxRandomSelectorV.Views;
```
이 프로젝트는 MVVM 패턴을 최대한 고수하려 하고 있습니다. Binding을 사용해주시기 바랍니다.
```suggestion
```

bequiett replied:
문제점을 인지하였으며 Views를 임포트하여 사용한 AlwaysOnTop의 핫키 코드를 MVVM 패턴을 위반하지 않는 방향으로 수정하겠습니다.

wowvv0w commented on DjmaxRandomSelectorV/ViewModels/MainViewModel.cs:
```diff
@@ -304,6 +304,25 @@ private IntPtr HwndHook(IntPtr hwnd, int msg, IntPtr wParam, IntPtr lParam, ref
+
+                        if (window.WindowState != WindowState.Minimized)
+                        {
+                            window.WindowState = WindowState.Minimized;
```
`Topmost`가 항상 `True`로 되어있어 불편함이 있습니다. 이 코드에 `Topmost` 값을 변경하는 코드를 넣는 것을 제안합니다.

bequiett replied:
MVVM 패턴 위반 오류 수정과 함께 변경하겠습니다.

wowvv0w commented on DjmaxRandomSelectorV/ViewModels/SettingViewModel.cs:
```diff
@@ -16,6 +19,7 @@ public class SettingViewModel : Screen
         private List<string> _ownedDlcs;
 
         private DockPanel _dockPanel;
+        private int updated;
```
좀 더 직관적인 변수명을 사용하면 좋을 것 같습니다.
```suggestion
private int dlcCount;
```

wowvv0w commented on DjmaxRandomSelectorV/Views/MainView.xaml:
```diff
@@ -3,29 +3,33 @@
-            <GradientStop Color="#2A2B2C" Offset="0.525"/>
-            <GradientStop Color="#2D2E2F" Offset="0.6"/>
-        </LinearGradientBrush>
+        <SolidColorBrush Opacity="0" Color="White" />
```
배경색을 `DockPanel.Background`로 옮긴 이유는 무엇입니까? `Window.Background`에 그대로 놔둬도 작동에 문제는 없었습니다.

bequiett replied:
전체화면 위에서 프로그램을 보이게 하기 위한 변경사항입니다.

wowvv0w replied:
```Window.Background```로 다시 옮기고 ```AllowsTranparency = true```를 지우고 시도해도 똑같이 작동하는 것을 확인했습니다. 따라서 ```Window.Background```로 다시 옮기는 것을 제안합니다. 혹시 놓친 의도가 있다면 말씀해주세요.
그리고 전체화면이라 함은 디맥 디스플레이 설정을 의미하는 것인가요? 전체화면 앱 위에 보이게 하려면 오버레이밖에 방법이 없는 것으로 알고 있습니다. 실제로 제가 디스플레이 설정을 Full Screen으로 하고 시도했을 때는 Alt+Tab과 다를 바 없이 작동했습니다.

bequiett replied:
AOT 기능은 Window의 Background가 투명이고 Topmost가 true로 설정되어 있을 때 정상적으로 동작합니다.
AOT 기능 추가와 함께 `Window.Background` 내 속성을 `Dockpanel.Background` 로 옮긴 것도 그 이유입니다.

제가 해당 기능을 시험하였을 때, Alt+Tab과 달리 디제이맥스 프로세스가 최소화 되지 않고 포커스만 풀렸고 랜덤선택기 프로세스가 최상단에 위치하였습니다.
변경된 코드는 디제이맥스 창에서 F2를 눌러 AOT를 동작시켰을 때 랜덤선택기 화면이 디제이맥스 프로세스 위에 뜨도록 설계되었으며 만약 의도대로 동작하지 않는다면 알려주세요. 정상 작동하지 않는 코드는 삭제하겠습니다.

wowvv0w replied:
실험 결과입니다:

----

1. VS2022의 '실행' 버튼을 통해 실행한 경우
> 디맥 디스플레이 설정: Bordered Window

(F2기능)ON: 포커스가 디맥에 고정됨, 정상 작동
OFF: 포커스가 디맥에 고정됨, 정상 작동

> 디맥 디스플레이 설정: Full Screen

ON: 포커스가 디맥에 고정되나 선곡기가 보이지 않음
*ON일 때 선곡기 창 크기만큼의 영역 위를 마우스 클릭하면 포커스가 풀리고 선곡기에 고정됨, 디맥 최소화됨
OFF: 포커스가 디맥에 고정됨, 정상 작동

----

2. exe를 직접 실행시킨 경우
> 디맥 디스플레이 설정: Bordered Window

ON: 포커스가 풀리고 선곡기에 고정됨, 디맥 최소화되지 않음
OFF: 포커스가 디맥에 고정됨, 정상 작동

> 디맥 디스플레이 설정: Full Screen

ON: 포커스가 풀리고 선곡기에 고정됨, 디맥 최소화됨, 
OFF: 다른 창(크롬 등)이 최소화 되어 있지 않을 시 해당 창으로 포커스가 고정됨, 모두 최소화 되어 있을 시 포커스가 디맥에 고정됨

----

어디서 실행하는가에 따라 결과가 달라지는게 이상하지만 원인을 모르겠습니다. 위의 결과와 동일하게 나타납니까?

bequiett replied:
해당 실험과 추가 실험을 진행하였습니다.

> DMRV 1.2.0 / Full Screen
정상 작동하지 않음

> DMRV 1.1.2 / Full Screen
이전에 정상작동함을 확인하였으나 현재 작동하지 않음

현재 해당 기능은 정상작동하지 않음을 확인하여 AOT 기능은 삭제하였습니다.

Commit `8b5c609` push 전 확인하였을 때 정상작동한 이유는 ALT+TAB 글리칭으로 인해 일시적으로 디제이맥스 프로세스가 보더리스 윈도우로 변경되어 작동한 것으로 보입니다.
또한 이전에 정상작동을 했던 것도 위와 관련이 있는것으로 보입니다.

wowvv0w replied:
Alt+Tab 글리치가 무엇인지와 DMRV 1.1.2 / 1.2.0이 무엇을 의미하는지에 대해 더 자세히 설명해주실 수 있으십니까? 디제이맥스의 버전은 v10xx를 쓰고 있는 것으로 알고 있습니다.

quiettbee replied:
해당 실험외에 추가 실험을 진행하였습니다.

> DMRV 1.2.0 / Full Screen
정상작동 X

> DMRV 1.1.2 / Full Screen
정상작동 X

현재 AOT 기능을 삭제하였습니다.
Commit [8b5c609](https://github.com/wowvv0w/djmax-random-selector-v/pull/4/commits/8b5c609083024ba6c5033e97508dc4f616e9a95b) 을 올리기 전에 한 실험에서 AOT 기능이 정상작동한 이유는 ALT+TAB 글리칭으로 인해 디제이맥스 프로세스가 보더리스 윈도우로 변경되어 정상작동한 것으로 보이며, 이전에 정상작동하였던 1.1.2 버전의 AOT가 작동하지 않는 이유는 다른 문제인것 같습니다.

bequiett replied:
해당 프로그램의 버전을 의미하나 제가 잘못 기재했습니다. 순서대로 1.1.0 베이스 [8b5c609](https://github.com/wowvv0w/djmax-random-selector-v/pull/4/commits/8b5c609083024ba6c5033e97508dc4f616e9a95b) 버전과 1.0.2 베이스인 [5b5e829](https://github.com/wowvv0w/djmax-random-selector-v/pull/4/commits/5b5e829b45f41a4afb78a40260a0f6b7efeacff8) 버전을 말합니다.

Alt+Tab 글리칭은 초반 로딩 시에나 VS에서 컴파일 진행 시에 프로그램에 렉이 걸린 경우 Alt+Tab 을 할 때 순간 프리징되며 Alt+Tab이 정상작동하지 않는 것을 말합니다. 이때 왜 인지는 모르겠지만 창의 모드가 변경되는 것으로 보입니다.
저같은 경우는 현재도 해당 증상이 보이며 프리징 이후 AOT를 사용하면 정상작동하였으며, 이 때 디제이맥스 설정에서 윈도우 모드 Full Screen을 다른 것으로 변경하였다가 다시 Full Screen으로 변경했을 때 AOT가 해제되었습니다.

wowvv0w commented on DjmaxRandomSelectorV/Views/SettingView.xaml:
```diff
@@ -23,15 +23,16 @@
                 <RowDefinition Height="20"/>
-                <RowDefinition/>
+                <RowDefinition Height="50*"/>
+                <RowDefinition Height="11*"/>
```
이 Row를 추가한 이유가 무엇입니까?

bequiett replied:
제거하였습니다.

wowvv0w commented on DjmaxRandomSelectorV/Views/SettingView.xaml:
```diff
@@ -60,15 +61,26 @@
+                               Grid.Column="0"/>
+                        <Button x:Name="Detect"
+                                Grid.Column="2"
+                                Style="{StaticResource SettingDetectButton}"/>
```
이런 유틸리티 기능을 하는 버튼은 `SettingWindowButton` 스타일을 사용합니다. `SettingWindowButton`을 사용해주십시오.

bequiett replied:
버튼 크기가 커 디자인을 해칠 우려가 있어 개별 스타일을 사용하였습니다.

wowvv0w commented on DjmaxRandomSelectorV/ViewModels/SettingViewModel.cs:
```diff
@@ -264,5 +290,32 @@ public bool SettingNXN
         }
+        public bool FindDLCs(string name)
+        {
+            bool dlccontained = true;
```
Boolean 명명규칙 적용
```suggestion
bool isContained = true;
```

wowvv0w reviewed (changes requested):
코멘트를 확인해주십시오.

DLC 감지 기능은 상상도 못했던 방식이었습니다. 이 기능 덕분에 사용하기 더 편리해질 것 같습니다. 감사합니다.

bequiett committed:
Make variable name intuitive

bequiett committed:
Make variable name intuitive 2

wowvv0w committed:
Merge branch 'main' into main

wowvv0w commented:
기능이 대거 추가됨에 따라 이 PR과 충돌이 다소 발생해, 두 커밋의 변경점을 유지한 채 해결하였습니다.

wowvv0w committed:
</style> 태그 누락 수정

Commit:
Fixes

Commit:
Delete not working feature.

The feature is only working when you use ALT+TAB glitch.

The feature is not properly working on the latest version of DJMAX and is deleted.
